File: run_microstates_lemon.py
"""
run_microstates_lemon.py - Python port of run_microstates_lemon.m

Replicates the MATLAB implementation exactly with same diagnostics.
"""

import sys
from pathlib import Path

import matplotlib
matplotlib.use("Agg")
import numpy as np
import matplotlib.pyplot as plt
import mne

# Add micro-VAR-states to path
sys.path.insert(0, "/home/kumarsak/micro-VAR-states/python")

from VAR import VAR
from microstates import (
    get_gfp_peaks,
    run_modified_k_means,
    KoenigTemplates,
    match_topomaps_per_template,
)

# =========================
# USER CONFIG
# =========================
LEMON_DIR = Path("/store/projects/kumarsak/LEMON_data")
OUT_DIR = Path("./microstate_results_python_var")

N_SUBJECTS = 1
N_STATES = 4
RANDOM_SEED = 42

np.random.seed(RANDOM_SEED)
OUT_DIR.mkdir(exist_ok=True)


# =========================
# UTILITIES
# =========================

def load_lemon_subjects(lemon_dir, n_subjects):
    """
    Load LEMON EEG .set files.

    Parameters
    ----------
    lemon_dir : Path
        Directory containing .set files
    n_subjects : int or str
        Number of subjects or 'all'

    Returns
    -------
    list of Path
        Sorted list of .set file paths
    """
    files = sorted(lemon_dir.glob("*_EC.set"))

    if isinstance(n_subjects, str) and n_subjects == 'all':
        return files

    if isinstance(n_subjects, int):
        return files[:min(n_subjects, len(files))]

    raise ValueError("n_subjects must be 'all' or an integer")


def subject_maps_to_raw(km_maps, info):
    """Convert microstate maps (states x channels) into MNE RawArray."""
    return mne.io.RawArray(km_maps.T, info, verbose=False)


def preprocess_for_gfp_peaks(raw, l_freq=1.0, h_freq=40.0, do_ica=False):
    """
    Classical EEGLAB-style preprocessing for microstates.

    Parameters
    ----------
    raw : mne.io.Raw
        MNE Raw object
    l_freq : float
        Low cutoff frequency
    h_freq : float
        High cutoff frequency
    do_ica : bool
        Whether to run ICA cleaning

    Returns
    -------
    gfp_peaks : np.ndarray
        Indices of GFP peaks
    gfp_curve : np.ndarray
        GFP time series
    raw : mne.io.Raw
        Preprocessed Raw object
    """
    # Band-pass filter (zero-phase FIR)
    raw = raw.copy().filter(
        l_freq=l_freq,
        h_freq=h_freq,
        method="fir",
        phase="zero",
        fir_design="firwin",
        verbose=False,
    )

    # Average reference (MANDATORY for microstates)
    raw = raw.set_eeg_reference("average", projection=False, verbose=False)

    # ICA cleaning (OPTIONAL)
    if do_ica:
        ica = mne.preprocessing.ICA(
            n_components=None,
            method="fastica",
            random_state=RANDOM_SEED,
            max_iter="auto",
        )
        ica.fit(raw, verbose=False)
        raw = ica.apply(raw.copy(), verbose=False)

    # GFP peaks on voltage data
    data = raw.get_data()
    gfp_peaks, gfp_curve = get_gfp_peaks(data)

    return gfp_peaks, gfp_curve, raw


def plot_microstate_maps(km_maps, info, subject_id, out_dir):
    """Plot and save microstate topographies."""
    fig, axes = plt.subplots(1, km_maps.shape[0], figsize=(3 * km_maps.shape[0], 3))

    if km_maps.shape[0] == 1:
        axes = [axes]

    for i, ax in enumerate(axes):
        mne.viz.plot_topomap(
            km_maps[i],
            info,
            axes=ax,
            show=False,
            contours=0,
            cmap="RdBu_r",
        )
        ax.set_title(f"State {chr(ord('A') + i)}")

    fig.suptitle(f"{subject_id} - Microstate Topographies")

    out_dir = Path(out_dir)
    out_dir.mkdir(exist_ok=True, parents=True)

    outpath = out_dir / f"{subject_id}_microstates_ABCD.png"
    fig.savefig(outpath, dpi=300, bbox_inches="tight")
    plt.close(fig)

    print(f"Saved microstate figure -> {outpath}")


def estimate_var_order(set_path):
    """
    Estimate VAR order for a subject.

    Parameters
    ----------
    set_path : Path
        Path to .set file

    Returns
    -------
    p_opt : int
        Optimal VAR order
    aic_vals : np.ndarray
        AIC values
    diagnostics : dict
        Order selection diagnostics
    """
    subj_id = set_path.stem
    print(f"\n=== Estimating VAR order for {subj_id} ===")

    # Load EEG
    raw = mne.io.read_raw_eeglab(set_path, preload=True, verbose=False)

    # Preprocess (same as main pipeline)
# --- VAR preprocessing (NO average reference) ---
    raw_var = raw.copy().filter(
        l_freq=1.0,
        h_freq=40.0,
        method="fir",
        phase="zero",
        fir_design="firwin",
        verbose=False,
    )

    data = raw_var.get_data()   # ❗ full-rank data



    # Create VAR model
    var_model = VAR(verbose=True, save_diagnostics=True)

    # Select order with p_max=10, elbow=True
    p_max = 10
    p_opt, aic_vals, diagnostics = var_model.select_order(data, p_max, elbow=True)

    return p_opt, aic_vals, diagnostics


def process_subject_fixed_p(set_path, out_dir_base, n_states, p_fixed):
    """
    Process subject with VAR-based microstates using fixed p.

    Matches MATLAB process_subject_fixed_p exactly.

    Parameters
    ----------
    set_path : Path
        Path to .set file
    out_dir_base : Path
        Base output directory
    n_states : int
        Number of microstates
    p_fixed : int
        Fixed VAR order

    Returns
    -------
    km_maps_labeled : np.ndarray
        Labeled microstate maps
    info : mne.Info
        MNE info object
    subj_id : str
        Subject ID
    out_dir : Path
        Subject output directory
    diagnostics : dict
        Full diagnostics
    """
    subj_id = set_path.stem
    print(f"\n=== Processing {subj_id} (p = {p_fixed}) ===")

    # Initialize diagnostics struct
    diagnostics = {
        'subject_id': subj_id,
        'p_fixed': p_fixed,
        'n_states': n_states
    }

    # --- Load EEG ---
    raw = mne.io.read_raw_eeglab(set_path, preload=True, verbose=False)

    diagnostics['original_data_size'] = raw.get_data().shape
    diagnostics['srate'] = raw.info['sfreq']

    # --- Preprocessing ---
    # --- VAR preprocessing (NO average reference) ---
    raw_var = raw.copy().filter(
        l_freq=1.0,
        h_freq=40.0,
        method="fir",
        phase="zero",
        fir_design="firwin",
        verbose=False,
    )

    data_var = raw_var.get_data()   # ❗ used for VAR

    # --- Microstate preprocessing (WITH average reference) ---
    raw_ms = raw.copy().filter(
        l_freq=1.0,
        h_freq=40.0,
        method="fir",
        phase="zero",
        fir_design="firwin",
        verbose=False,
    )
    raw_ms.set_eeg_reference("average", projection=False, verbose=False)


    diagnostics['preprocessed_data_size'] = data_var.shape
    diagnostics['data_range_original'] = [float(np.min(data_var)), float(np.max(data_var))]

    # --- VAR with fixed order ---
    var_model = VAR(verbose=True, save_diagnostics=True)
    model = var_model.fit(data_var, p_fixed)

    # Store fit diagnostics
    if 'diagnostics' in model:
        diagnostics['var_fit'] = model['diagnostics']

    # --- Simulate long VAR data ---
    Tvar = int(3600 * raw.info['sfreq'])  # 1 hour of data
    data_sim, sim_diag = var_model.simulate(model, Tvar)

    # Store simulation diagnostics
    diagnostics['var_simulation'] = sim_diag

    # Detailed checks
    diagnostics['simulation_checks'] = {
        'size': data_sim.shape,
        'range': [float(np.min(data_sim)), float(np.max(data_sim))],
        'contains_nan': bool(np.any(np.isnan(data_sim))),
        'contains_inf': bool(np.any(np.isinf(data_sim))),
        'mean_abs': float(np.mean(np.abs(data_sim))),
        'std_mean': float(np.mean(np.std(data_sim, axis=1)))
    }

    # ABORT if simulation failed
    if (diagnostics['simulation_checks']['contains_nan'] or
        diagnostics['simulation_checks']['contains_inf'] or
        diagnostics['simulation_checks']['mean_abs'] > 1e6):
        raise RuntimeError("VAR simulation produced invalid data. Model is still unstable.")
    # --- Apply average reference to simulated data (for microstates) ---
    data_sim = data_sim - data_sim.mean(axis=0, keepdims=True)

    # --- GFP Analysis ---
    gfp_peaks_var, gfp_curve = get_gfp_peaks(data_sim)

    diagnostics['gfp'] = {
        'n_peaks': len(gfp_peaks_var),
        'peak_range': [int(np.min(gfp_peaks_var)), int(np.max(gfp_peaks_var))],
        'gfp_curve_range': [float(np.min(gfp_curve)), float(np.max(gfp_curve))],
        'gfp_curve_mean': float(np.mean(gfp_curve))
    }

    print(f"\n=== GFP Analysis ===")
    print(f"Number of GFP peaks: {len(gfp_peaks_var)}")
    print(f"GFP curve range: [{np.min(gfp_curve):.4e}, {np.max(gfp_curve):.4e}]")

    if len(gfp_peaks_var) < n_states:
        raise RuntimeError(f"Not enough GFP peaks ({len(gfp_peaks_var)}) for {n_states} states")

    peaks_data = data_sim[:, gfp_peaks_var]
    diagnostics['gfp']['peaks_data_range'] = [float(np.min(peaks_data)), float(np.max(peaks_data))]
    diagnostics['gfp']['peaks_contain_nan'] = bool(np.any(np.isnan(peaks_data)))

    # --- Microstates ---
    km_maps = run_modified_k_means(data_sim[:, gfp_peaks_var], n_states=n_states)

    diagnostics['microstates'] = {
        'n_states': km_maps.shape[0],
        'n_channels': km_maps.shape[1],
        'maps_range': [float(np.min(km_maps)), float(np.max(km_maps))]
    }

    # --- Template matching ---
    subject_maps_raw = subject_maps_to_raw(km_maps, raw.info)
    templates = KoenigTemplates.load(n_states=n_states)

    attribution, corr = match_topomaps_per_template(
        maps=subject_maps_raw,
        template=templates
    )

    km_maps_labeled = km_maps[attribution, :]

    diagnostics['template_matching'] = {
        'attribution': list(attribution),
        'correlations': list(corr),
        'mean_correlation': float(np.mean(corr))
    }

    # --- Save everything ---
    out_dir = Path(out_dir_base) / subj_id
    out_dir.mkdir(exist_ok=True, parents=True)

    # Save model and results
    np.save(out_dir / "var_model.npy", model, allow_pickle=True)
    np.save(out_dir / "var_p_fixed.npy", p_fixed)
    np.save(out_dir / "var_km_maps.npy", km_maps)
    np.save(out_dir / "var_km_maps_labeled.npy", km_maps_labeled)
    np.save(out_dir / "var_permutation.npy", attribution)

    # Save comprehensive diagnostics
    np.save(out_dir / "var_diagnostics_full.npy", diagnostics, allow_pickle=True)

    # Also save a summary text file
    write_diagnostics_summary(out_dir / "var_diagnostics_summary.txt", diagnostics)

    return km_maps_labeled, raw.info, subj_id, out_dir, diagnostics


def write_diagnostics_summary(filename, diagnostics):
    """Write human-readable diagnostics summary."""
    with open(filename, 'w') as f:
        f.write("=== VAR-Microstate Analysis Diagnostics ===\n\n")
        f.write(f"Subject: {diagnostics['subject_id']}\n")
        f.write(f"VAR Order: {diagnostics['p_fixed']}\n")
        f.write(f"Number of States: {diagnostics['n_states']}\n\n")

        f.write("--- Data ---\n")
        f.write(f"Original size: {diagnostics['original_data_size'][0]} x {diagnostics['original_data_size'][1]}\n")
        f.write(f"Sampling rate: {diagnostics['srate']:.1f} Hz\n")
        f.write(f"Data range: [{diagnostics['data_range_original'][0]:.4e}, {diagnostics['data_range_original'][1]:.4e}]\n\n")

        if 'var_fit' in diagnostics:
            f.write("--- VAR Fitting ---\n")
            vf = diagnostics['var_fit']
            f.write(f"Original max eigenvalue: {vf['max_eig_original']:.6f}\n")
            f.write(f"Final max eigenvalue: {vf['max_eig_final']:.6f}\n")
            f.write(f"Rescale iterations: {vf['rescale_iterations']}\n")
            f.write(f"Stabilized: {vf.get('stabilized', False)}\n\n")

        if 'var_simulation' in diagnostics:
            f.write("--- VAR Simulation ---\n")
            vs = diagnostics['var_simulation']
            f.write(f"Simulated length: {vs['T']} samples\n")
            f.write(f"Sigma condition number: {vs['sigma_condition']:.4e}\n")
            f.write(f"Data range: [{vs['data_range'][0]:.4e}, {vs['data_range'][1]:.4e}]\n")
            f.write(f"Contains NaN: {vs['contains_nan']}\n")
            f.write(f"Contains Inf: {vs['contains_inf']}\n")
            f.write(f"Mean absolute value: {vs['mean_abs']:.4e}\n")
            f.write(f"Mean std: {vs['mean_std']:.4e}\n\n")

        if 'gfp' in diagnostics:
            f.write("--- GFP Analysis ---\n")
            gfp = diagnostics['gfp']
            f.write(f"Number of peaks: {gfp['n_peaks']}\n")
            f.write(f"GFP curve range: [{gfp['gfp_curve_range'][0]:.4e}, {gfp['gfp_curve_range'][1]:.4e}]\n")
            f.write(f"Peaks data range: [{gfp['peaks_data_range'][0]:.4e}, {gfp['peaks_data_range'][1]:.4e}]\n\n")

        if 'template_matching' in diagnostics:
            f.write("--- Template Matching ---\n")
            tm = diagnostics['template_matching']
            f.write(f"Attribution: {tm['attribution']}\n")
            f.write(f"Correlations: {[f'{c:.4f}' for c in tm['correlations']]}\n")
            f.write(f"Mean correlation: {tm['mean_correlation']:.4f}\n")

    print(f"Diagnostics summary saved to: {filename}")


# =========================
# MAIN
# =========================

if __name__ == "__main__":
    # Load subjects
    sets = load_lemon_subjects(LEMON_DIR, N_SUBJECTS)
    nSubj = len(sets)

    # ---------- PASS 1: estimate p for each subject ----------
    print("\n=== Estimating VAR order per subject ===")

    p_opts = []

    for i, set_path in enumerate(sets):
        # Get p_opt, AIC values, AND diagnostics
        p_opt, aic_vals, order_diagnostics = estimate_var_order(set_path)
        p_opts.append(p_opt)

        # Get subject ID
        subj_id = set_path.stem

        # Create output directory
        out_dir = OUT_DIR / subj_id
        out_dir.mkdir(exist_ok=True, parents=True)

        # Save AIC values, optimal p, AND diagnostics
        np.save(out_dir / "var_aic.npy", aic_vals)
        np.save(out_dir / "var_p_opt.npy", p_opt)
        np.save(out_dir / "var_order_selection_diagnostics.npy", order_diagnostics, allow_pickle=True)

        print(f"Subject {i + 1} ({subj_id}): p_opt = {p_opt}")

    p_med = int(np.median(p_opts))
    print(f"\n>>> Median VAR order across subjects: p = {p_med} <<<")

    np.save(OUT_DIR / "var_popts_all_subjects.npy", p_opts)
    np.save(OUT_DIR / "var_p_median.npy", p_med)

    # ---------- PASS 2: full pipeline using fixed p ----------
    print("\n=== Running full VAR + microstates with fixed p ===")

    all_diagnostics = []

    for i, set_path in enumerate(sets):
        km_maps, info, subj_id, out_dir, diagnostics = process_subject_fixed_p(
            set_path, OUT_DIR, N_STATES, p_med
        )

        plot_microstate_maps(km_maps, info, subj_id, out_dir)

        # Store diagnostics for aggregation
        all_diagnostics.append(diagnostics)

    # Save aggregated diagnostics across all subjects
    np.save(OUT_DIR / "all_subjects_diagnostics.npy", all_diagnostics, allow_pickle=True)

    print("\nAll subjects processed.")

File: run_microstates_lemon.m
%% =========================
% ENVIRONMENT SETUP
% =========================

addpath('/home/kumarsak/micro-VAR-states/matlab');

rng(42);
% --- EEGLAB + SIFT ---
addpath('/home/kumarsak/eeglab2025.1.0');
eeglab nogui;

addpath(genpath('/home/kumarsak/eeglab2025.1.0/plugins/SIFT'));
% addpath(genpath('/home/kumarsak/VAR-Toolbox/v3dot0/'))

LEMON_DIR    = '/store/projects/kumarsak/LEMON_data';
OUT_DIR      = './microstate_results_matlab_var';

N_SUBJECTS = 1;
N_STATES   = 4;
SUBJECT_ID =  'sub-010006_EC'; % [] means random


if ~exist(OUT_DIR, 'dir')
    mkdir(OUT_DIR);
end

function sets = load_lemon_subjects(lemon_dir, n_subjects)
%LOAD_LEMON_SUBJECTS Load LEMON EEG .set files
%
% n_subjects:
%   'all' → all subjects
%   integer → first n subjects

    files = dir(fullfile(lemon_dir, '*_EC.set'));

    % sort deterministically
    [~, idx] = sort({files.name});
    files = files(idx);

    if ischar(n_subjects) || isstring(n_subjects)
        if strcmp(n_subjects, 'all')
            sets = files;
            return
        else
            error("n_subjects must be 'all' or an integer");
        end
    end

    if isnumeric(n_subjects) && isscalar(n_subjects)
        sets = files(1:min(n_subjects, numel(files)));
        return
    end

    error("n_subjects must be 'all' or an integer");
end

function subject_maps = subject_maps_to_struct(km_maps, EEG)
    subject_maps.data     = km_maps;     % states × channels
    subject_maps.chanlocs = EEG.chanlocs;
    subject_maps.nbchan   = EEG.nbchan;
    subject_maps.n_states = size(km_maps, 1);
end
function [gfp_peaks, gfp_curve, EEG] = preprocess_for_gfp_peaks(EEG, l_freq, h_freq, do_ica)
%PREPROCESS_FOR_GFP_PEAKS Classical EEGLAB microstate preprocessing
%
% Inputs
%   EEG     : EEGLAB EEG structure
%   l_freq  : low cutoff frequency (e.g. 1)
%   h_freq  : high cutoff frequency (e.g. 40)
%   do_ica  : true / false (whether to run ICA cleaning)
%
% Outputs
%   gfp_peaks : indices of GFP peaks
%   gfp_curve : GFP time series
%   EEG       : cleaned EEG (important!)

    if nargin < 4
        do_ica = false;
    end

    %% --------------------------------------------------
    % 1. Band-pass filter (EEGLAB FIR, zero-phase)
    %% -------------------------------------------------
    EEG = pop_eegfiltnew(EEG, l_freq, h_freq);
    EEG = eeg_checkset(EEG);
    %% --------------------------------------------------
    % 3. Average reference (MANDATORY for microstates)
    %% --------------------------------------------------
    EEG = pop_reref(EEG, []);
    EEG = eeg_checkset(EEG);

    %% --------------------------------------------------
    % 2. Run ICA for artifact removal (OPTIONAL but recommended)
    %% --------------------------------------------------
    if do_ica
        % Run ICA
        EEG = pop_runica(EEG, 'extended', 1, 'stop', 1e-7);
        EEG = eeg_checkset(EEG);

        % Classify components
        EEG = pop_iclabel(EEG, 'default');

        % Reject non-brain components automatically
        EEG = pop_icflag(EEG, ...
            [ NaN NaN;    % Brain
              0.9 1;      % Muscle
              0.9 1;      % Eye
              0.9 1;      % Heart
              0.9 1;      % Line Noise
              0.9 1;      % Channel Noise
              NaN NaN ]); % Other

        EEG = eeg_checkset(EEG);
    end
    %% --------------------------------------------------
    % 4. GFP peaks on voltage data
    %% --------------------------------------------------
    data = double(EEG.data);   % channels × time

    ms = microVARstates.microstates();
    [gfp_peaks, gfp_curve] = ms.get_gfp_peaks(data);

end

function plot_microstate_maps(km_maps, EEG, subject_id, out_dir)

    figure('Visible','off');
    n_states = size(km_maps,1);

    for s = 1:n_states
        subplot(1,n_states,s);
        topoplot(km_maps(s,:), EEG.chanlocs, ...
                 'electrodes','off');
        title(['State ' char('A'+s-1)]);
    end

    sgtitle([subject_id ' – Microstate Topographies']);

    if ~exist(out_dir,'dir')
        mkdir(out_dir);
    end

    saveas(gcf, fullfile(out_dir, ...
        [subject_id '_microstates_ABCD.png']));
    close;
end

function [p_opt, aic_vals, diagnostics] = estimate_var_order(setfile)
    EEG = pop_loadset('filename', setfile.name, ...
                      'filepath', setfile.folder);
    
    % Same preprocessing as main pipeline
    [~, ~, EEG] = preprocess_for_gfp_peaks(EEG, 1, 40, false);
    data = double(EEG.data);
    
    varModel = microVARstates.VAR();
    varModel.verbose = true;  % Print diagnostics
    varModel.save_diagnostics = true;  % Include in output
    
    % IMPORTANT: use limited p_max
    p_max = 10;
    [p_opt, aic_vals, diagnostics] = varModel.select_order(data, p_max, true);
end

function [km_maps_labeled, EEG, subj_id, out_dir, diagnostics] = ...
    process_subject_fixed_p(setfile, OUT_DIR, N_STATES, p_fixed)

    subj_id = erase(setfile.name, '.set');
    fprintf('\n=== Processing %s (p = %d) ===\n', subj_id, p_fixed);
    
    % Initialize diagnostics struct
    diagnostics = struct();
    diagnostics.subject_id = subj_id;
    diagnostics.p_fixed = p_fixed;
    diagnostics.n_states = N_STATES;

    %% --- Load EEG ---
    EEG = pop_loadset('filename', setfile.name, ...
                      'filepath', setfile.folder);
    
    diagnostics.original_data_size = size(EEG.data);
    diagnostics.srate = EEG.srate;

    %% --- Preprocessing ---
    [~, ~, EEG] = preprocess_for_gfp_peaks(EEG, 1, 40, false);
    data = double(EEG.data);
    
    diagnostics.preprocessed_data_size = size(data);
    diagnostics.data_range_original = [min(data(:)), max(data(:))];

    %% --- VAR with fixed order ---
    varModel = microVARstates.VAR();
    varModel.verbose = true;
    varModel.save_diagnostics = true;

    model = varModel.fit(data, p_fixed);
    
    % Store fit diagnostics
    if isfield(model, 'diagnostics')
        diagnostics.var_fit = model.diagnostics;
    end

    %% --- Simulate long VAR data ---
    Tvar = 3600 * EEG.srate;
    [data_sim, sim_diag] = varModel.simulate(model, Tvar);
    
    % Store simulation diagnostics
    diagnostics.var_simulation = sim_diag;
    
    % Detailed checks
    diagnostics.simulation_checks = struct();
    diagnostics.simulation_checks.size = size(data_sim);
    diagnostics.simulation_checks.range = [min(data_sim(:)), max(data_sim(:))];
    diagnostics.simulation_checks.contains_nan = any(isnan(data_sim(:)));
    diagnostics.simulation_checks.contains_inf = any(isinf(data_sim(:)));
    diagnostics.simulation_checks.mean_abs = mean(abs(data_sim(:)));
    diagnostics.simulation_checks.std_mean = mean(std(data_sim, 0, 2));
    
    % ABORT if simulation failed
    if diagnostics.simulation_checks.contains_nan || ...
       diagnostics.simulation_checks.contains_inf || ...
       diagnostics.simulation_checks.mean_abs > 1e6
        error('VAR simulation produced invalid data. Model is still unstable.');
    end

    %% --- GFP Analysis ---
    ms = microVARstates.microstates();
    [gfp_peaks_var, gfp_curve] = ms.get_gfp_peaks(data_sim);
    
    diagnostics.gfp = struct();
    diagnostics.gfp.n_peaks = length(gfp_peaks_var);
    diagnostics.gfp.peak_range = [min(gfp_peaks_var), max(gfp_peaks_var)];
    diagnostics.gfp.gfp_curve_range = [min(gfp_curve), max(gfp_curve)];
    diagnostics.gfp.gfp_curve_mean = mean(gfp_curve);
    
    if length(gfp_peaks_var) < N_STATES
        error('Not enough GFP peaks (%d) for %d states', ...
              length(gfp_peaks_var), N_STATES);
    end
    
    peaks_data = data_sim(:, gfp_peaks_var);
    diagnostics.gfp.peaks_data_range = [min(peaks_data(:)), max(peaks_data(:))];
    diagnostics.gfp.peaks_contain_nan = any(isnan(peaks_data(:)));

    %% --- Microstates ---
    km_maps = ms.run_modified_k_means(data_sim(:, gfp_peaks_var), N_STATES);
    
    diagnostics.microstates = struct();
    diagnostics.microstates.n_states = size(km_maps, 1);
    diagnostics.microstates.n_channels = size(km_maps, 2);
    diagnostics.microstates.maps_range = [min(km_maps(:)), max(km_maps(:))];

    %% --- Template matching ---
    subject_maps = subject_maps_to_struct(km_maps, EEG);
    templates = microVARstates.KoenigTemplates.load(N_STATES);

    [attribution, corr] = ms.match_topomaps_per_template( ...
        subject_maps.data.', subject_maps.chanlocs, templates);

    km_maps_labeled = km_maps(attribution, :);
    
    diagnostics.template_matching = struct();
    diagnostics.template_matching.attribution = attribution;
    diagnostics.template_matching.correlations = corr;
    diagnostics.template_matching.mean_correlation = mean(corr);

    %% --- Save everything ---
    out_dir = fullfile(OUT_DIR, subj_id);
    if ~exist(out_dir,'dir')
        mkdir(out_dir);
    end
    
    % Save model and results
    save(fullfile(out_dir,'var_model.mat'), 'model');
    save(fullfile(out_dir,'var_p_fixed.mat'), 'p_fixed');
    save(fullfile(out_dir,'var_km_maps.mat'), 'km_maps');
    save(fullfile(out_dir,'var_km_maps_labeled.mat'), 'km_maps_labeled');
    save(fullfile(out_dir,'var_permutation.mat'), 'attribution');
    
    % Save comprehensive diagnostics
    save(fullfile(out_dir,'var_diagnostics_full.mat'), 'diagnostics');
    
    % Also save a summary text file
    write_diagnostics_summary(fullfile(out_dir,'var_diagnostics_summary.txt'), diagnostics);
end

function write_diagnostics_summary(filename, diagnostics)
    % Write human-readable diagnostics summary
    fid = fopen(filename, 'w');
    
    fprintf(fid, '=== VAR-Microstate Analysis Diagnostics ===\n\n');
    fprintf(fid, 'Subject: %s\n', diagnostics.subject_id);
    fprintf(fid, 'VAR Order: %d\n', diagnostics.p_fixed);
    fprintf(fid, 'Number of States: %d\n\n', diagnostics.n_states);
    
    fprintf(fid, '--- Data ---\n');
    fprintf(fid, 'Original size: %d × %d\n', diagnostics.original_data_size(1), diagnostics.original_data_size(2));
    fprintf(fid, 'Sampling rate: %.1f Hz\n', diagnostics.srate);
    fprintf(fid, 'Data range: [%.4e, %.4e]\n\n', diagnostics.data_range_original(1), diagnostics.data_range_original(2));
    
    if isfield(diagnostics, 'var_fit')
        fprintf(fid, '--- VAR Fitting ---\n');
        fprintf(fid, 'Original max eigenvalue: %.6f\n', diagnostics.var_fit.max_eig_original);
        fprintf(fid, 'Final max eigenvalue: %.6f\n', diagnostics.var_fit.max_eig_final);
        fprintf(fid, 'Rescale iterations: %d\n', diagnostics.var_fit.rescale_iterations);
        if isfield(diagnostics.var_fit, 'stabilized')
            fprintf(fid, 'Stabilized: %d\n\n', diagnostics.var_fit.stabilized);
        else
            fprintf(fid, 'Stabilized: 0\n\n');
        end

    end
    
    if isfield(diagnostics, 'var_simulation')
        fprintf(fid, '--- VAR Simulation ---\n');
        fprintf(fid, 'Simulated length: %d samples\n', diagnostics.var_simulation.T);
        fprintf(fid, 'Sigma condition number: %.4e\n', diagnostics.var_simulation.sigma_condition);
        fprintf(fid, 'Data range: [%.4e, %.4e]\n', diagnostics.var_simulation.data_range(1), diagnostics.var_simulation.data_range(2));
        fprintf(fid, 'Contains NaN: %d\n', diagnostics.var_simulation.contains_nan);
        fprintf(fid, 'Contains Inf: %d\n', diagnostics.var_simulation.contains_inf);
        fprintf(fid, 'Mean absolute value: %.4e\n', diagnostics.var_simulation.mean_abs);
        fprintf(fid, 'Mean std: %.4e\n\n', diagnostics.var_simulation.mean_std);
    end
    
    if isfield(diagnostics, 'gfp')
        fprintf(fid, '--- GFP Analysis ---\n');
        fprintf(fid, 'Number of peaks: %d\n', diagnostics.gfp.n_peaks);
        fprintf(fid, 'GFP curve range: [%.4e, %.4e]\n', diagnostics.gfp.gfp_curve_range(1), diagnostics.gfp.gfp_curve_range(2));
        fprintf(fid, 'Peaks data range: [%.4e, %.4e]\n\n', diagnostics.gfp.peaks_data_range(1), diagnostics.gfp.peaks_data_range(2));
    end
    
    if isfield(diagnostics, 'template_matching')
        fprintf(fid, '--- Template Matching ---\n');
        fprintf(fid, 'Attribution: [%s]\n', num2str(diagnostics.template_matching.attribution));
        fprintf(fid, 'Correlations: [%s]\n', num2str(diagnostics.template_matching.correlations, '%.4f '));
        fprintf(fid, 'Mean correlation: %.4f\n', diagnostics.template_matching.mean_correlation);
    end
    
    fclose(fid);
    fprintf('Diagnostics summary saved to: %s\n', filename);
end

%% =========================
% MODIFIED LOOP IN MAIN SCRIPT
% =========================

% In your main script, update the loops like this:
%% =========================
% LOAD SUBJECTS (THIS WAS MISSING)
% =========================

if ~isempty(SUBJECT_ID)
    % Load exactly one subject
    sets = dir(fullfile(LEMON_DIR, [SUBJECT_ID '.set']));
    if isempty(sets)
        error('Subject not found: %s', SUBJECT_ID);
    end
else
    % Load first N subjects
    sets = load_lemon_subjects(LEMON_DIR, N_SUBJECTS);
end

nSubj = numel(sets);


%% ---------- PASS 1: estimate p for each subject ----------
fprintf('\n=== Estimating VAR order per subject ===\n');

for i = 1:nSubj
    % Get p_opt, AIC values, AND diagnostics
    [p_opts(i), aic_vals, order_diagnostics] = estimate_var_order(sets(i));
    
    % Get subject ID
    subj_id = erase(sets(i).name, '.set');
    
    % Create output directory
    out_dir = fullfile(OUT_DIR, subj_id);
    if ~exist(out_dir, 'dir')
        mkdir(out_dir);
    end
    
    % Save AIC values, optimal p, AND diagnostics
    p_opt = p_opts(i);
    save(fullfile(out_dir, 'var_aic.mat'), 'aic_vals', 'p_opt');
    save(fullfile(out_dir, 'var_order_selection_diagnostics.mat'), 'order_diagnostics');
    
    fprintf('Subject %d (%s): p_opt = %d\n', i, subj_id, p_opts(i));
end

p_med = median(p_opts);
fprintf('\n>>> Median VAR order across subjects: p = %d <<<\n', p_med);

save(fullfile(OUT_DIR,'var_popts_all_subjects.mat'), 'p_opts', 'p_med');

%% ---------- PASS 2: full pipeline using fixed p ----------
fprintf('\n=== Running full VAR + microstates with fixed p ===\n');

all_diagnostics = cell(nSubj, 1);

for i = 1:nSubj
    [km_maps, EEG, subj_id, out_dir, diagnostics] = ...
        process_subject_fixed_p(sets(i), OUT_DIR, N_STATES, 10);

    plot_microstate_maps(km_maps, EEG, subj_id, out_dir);
    
    % Store diagnostics for aggregation
    all_diagnostics{i} = diagnostics;
end

% Save aggregated diagnostics across all subjects
save(fullfile(OUT_DIR, 'all_subjects_diagnostics.mat'), 'all_diagnostics');

disp('All subjects processed.');
